File: modules/core/app/aizou/core/GeoAPI.java
package aizou.core;

import com.lvxingpai.yunkai.UserInfo;
import exception.AizouException;
import models.AizouBaseEntity;
import database.MorphiaFactory;
import models.geo.Country;
import models.geo.Locality;
import models.misc.Track;
import org.bson.types.ObjectId;
import org.mongodb.morphia.Datastore;
import org.mongodb.morphia.query.CriteriaContainerImpl;
import org.mongodb.morphia.query.Query;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Created by zephyre on 7/10/14.
 */
public class GeoAPI {

    /**
     * 获得国家详情
     *
     * @param countryId
     * @param field
     * @return
     */
    public static Country countryDetails(ObjectId countryId, List<String> field) throws AizouException {
        Datastore ds = MorphiaFactory.datastore();
        Query<Country> query = ds.createQuery(Country.class).field("_id").equal(countryId);
        if (field != null && !field.isEmpty())
            query.retrievedFields(true, field.toArray(new String[field.size()]));

        return query.get();
    }

    public static Country countryDetails(String countryId) throws AizouException {
        return countryDetails(new ObjectId(countryId), null);
    }

    public static List<Country> getAllCountryList(List<String> field) throws AizouException {
        Datastore ds = MorphiaFactory.datastore();
        Query<Country> query = ds.createQuery(Country.class);
        if (field != null && !field.isEmpty())
            query.retrievedFields(true, field.toArray(new String[field.size()]));
        return query.asList();
    }

    public static List<Country> getAllCountryList() throws AizouException {
        return getAllCountryList(null);
    }

    // 后续通过配置做，目前只是写一个接口
    public static List<UserInfo> getExpertsByCountry(String countryId) throws AizouException {
        Datastore ds = MorphiaFactory.datastore();
        // 通过国家ID找到所有的足迹
        Query<Track> query = ds.createQuery(Track.class);
        // 得到所有足迹后，找到userId
        // 根据userId找到达人信息
        List<UserInfo> experts = null;
        return experts;
    }

    /**
     * 获得城市详情。
     *
     * @param locId 城市ID。
     * @return 如果没有找到，返回null。
     * @throws exception.AizouException
     */
    public static Locality locDetails(ObjectId locId, List<String> fields) throws AizouException {
        Datastore ds = MorphiaFactory.datastore();
        Query<Locality> query = ds.createQuery(Locality.class).field("_id").equal(locId).field(AizouBaseEntity.FD_TAOZIENA).equal(true);
        if (fields != null && !fields.isEmpty())
            query.retrievedFields(true, fields.toArray(new String[fields.size()]));
        return query.get();
    }

    /**
     * 获得城市详情。
     *
     * @param locId 城市ID。
     * @return 如果没有找到，返回null。
     * @throws exception.AizouException
     */
    public static Locality locDetails(String locId) throws AizouException {
        return locDetails(new ObjectId(locId), null);
    }

    /**
     * 通过关键词对目的地进行搜索。
     *
     * @param keyword  搜索关键词。
     * @param prefix   是否为前缀搜索？
     * @param page     分页偏移量。
     * @param pageSize 页面大小。
     */
    public static java.util.Iterator<Locality> searchLocalities(String keyword, boolean prefix, ObjectId countryId,
                                                                int page, int pageSize) throws AizouException {
        return searchLocalities(keyword, prefix, countryId, page, pageSize, null);
    }

    public static java.util.Iterator<Locality> searchLocalities(String keyword, boolean prefix, ObjectId countryId,
                                                                int page, int pageSize, Collection<String> fields)
            throws AizouException {
        Datastore ds = MorphiaFactory.datastore();
        Query<Locality> query = ds.createQuery(Locality.class);
        if (keyword != null && !keyword.isEmpty())
            query.field("alias").equal(Pattern.compile("^" + keyword));

        if (countryId != null)
            query.field(String.format("%s.id", Locality.fnCountry)).equal(countryId);
        query.field(AizouBaseEntity.FD_TAOZIENA).equal(true);

        if (fields != null && !fields.isEmpty())
            query.retrievedFields(true, fields.toArray(new String[fields.size()]));
        return query.order(String.format("-%s", Locality.fnHotness))
                .offset(page * pageSize).limit(pageSize).iterator();
    }

    /**
     * 根据名称搜索国家。
     *
     * @param keyword
     * @return
     * @throws exception.AizouException
     */
    public static List<Country> searchCountryByName(String keyword, int page, int pageSize) throws AizouException {
        Query<Country> query = MorphiaFactory.datastore().createQuery(Country.class);
        if (!keyword.equals("")) {
            query.field("alias").equal(Pattern.compile("^" + keyword));
        }
        query.offset(page * pageSize).limit(pageSize);
        return query.asList();
    }

    /**
     * 根据名称搜索国家。
     *
     * @return
     * @throws exception.AizouException
     */
    public static List<Country> searchCountryByName(List<String> keywords, int page, int pageSize) throws AizouException {
        Query<Country> query = MorphiaFactory.datastore().createQuery(Country.class);
        if (keywords != null) {
            List<CriteriaContainerImpl> criList = new ArrayList<>();
            for (String word : keywords)
                criList.add(query.criteria("alias").equal(Pattern.compile("^" + word)));
            query.or(criList.toArray(new CriteriaContainerImpl[criList.size()]));
        }
        query.field(AizouBaseEntity.FD_TAOZIENA).equal(true);
        query.offset(page * pageSize).limit(pageSize);
        return query.asList();
    }

    /**
     * 搜索目的地
     *
     * @param
     * @param page
     * @param pageSize
     * @return
     */
    public static List<Locality> getDestinations(boolean abroad, int page, int pageSize) throws AizouException {

        Datastore ds = MorphiaFactory.datastore();
        Query<Locality> query = ds.createQuery(Locality.class);
        query.field("abroad").equal(abroad).field("enabled").equal(true);
        query.offset(page * pageSize).limit(pageSize);
        query.order("-hotness");
        return query.asList();
    }

    public static List<Locality> getDestinationsByCountry(ObjectId countryID, int page, int pageSize) throws AizouException {
        Datastore ds = MorphiaFactory.datastore();
        Query<Locality> query = ds.createQuery(Locality.class);
        query.field("country._id").equal(countryID).field(AizouBaseEntity.FD_TAOZIENA).equal(true);
        query.order("-hotness");
        query.offset(page * pageSize).limit(pageSize);
        query.retrievedFields(true, AizouBaseEntity.FD_ID, Locality.FD_ZH_NAME, Locality.FD_EN_NAME, Locality.fnLocation, Locality.fnImages);
        return query.asList();
    }
}

File: modules/core/app/api/GeoAPI.scala
package api

import java.util

import com.twitter.util.{ Future, FuturePool }
import models.geo.{ Country, CountryExpert, Locality }
import org.bson.types.ObjectId
import org.mongodb.morphia.Datastore

import scala.collection.JavaConversions._
import scala.util.Random
import scala.language.postfixOps

/**
 * Created by topy on 2015/7/23.
 */
object GeoAPI {

  def getCountryExperts()(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[CountryExpert]] = {
    futurePool {
      val query = ds.createQuery(classOf[CountryExpert])
      query.asList() map (c => {
        val images = c.getImages
        if (images != null && images.size() > 1) {
          val idx = Random.nextInt(images.size())
          c.setImages(util.Arrays.asList(images.get(idx)))
        }
        c
      })
    }
  }

  def getCountryRecommend(fields: Seq[String] = Seq())(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Country]] = {
    futurePool {
      val query = ds.createQuery(classOf[Country])
        .field("isHot").equal(true)
        .field("enabled").equal(true)
        .order("rank")
      (if (fields nonEmpty) query.retrievedFields(true, fields: _*) else query).asList()
    }
  }

  def getCountryByNames(names: Seq[String], fields: Seq[String] = Seq())(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Country]] = {
    futurePool {
      if (names == null || names.isEmpty)
        Seq()
      else {
        val query = ds.createQuery(classOf[Country]).field(Country.FD_ZH_NAME).in(names) //.retrievedFields(true, fields: _*)
        (if (fields nonEmpty) query.retrievedFields(true, fields: _*) else query).asList()
      }
    }
  }

  def getCountryByIds(ids: Seq[ObjectId])(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Country]] = {
    futurePool {
      if (ids == null || ids.isEmpty)
        Seq()
      else {
        val query = ds.createQuery(classOf[Country]).field(Country.FN_ID).in(ids).retrievedFields(true, Seq(Country.FD_ZH_NAME): _*)
        query.asList()
      }
    }
  }

  def getCountryByContCode(contCode: String, fields: Seq[String] = Seq())(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Country]] = {
    futurePool {
      if (contCode == null)
        Seq()
      else {
        val query = ds.createQuery(classOf[Country])
          .field(Country.fnContCode).equal(contCode)
          .field("enabled").equal(true)
          .order("rank")
        (if (fields nonEmpty) query.retrievedFields(true, fields: _*) else query).asList()
      }
    }
  }

  def getLocalityByNames(names: Seq[String], fields: Seq[String] = Seq())(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Locality]] = {
    futurePool {
      if (names == null || names.isEmpty)
        Seq()
      else {
        val query = ds.createQuery(classOf[Locality]).field(Locality.FD_ZH_NAME).in(names) //.retrievedFields(true, fields: _*)
        (if (fields nonEmpty) query.retrievedFields(true, fields: _*) else query).asList()
      }
    }
  }

  def getLocalityById(id: ObjectId, fields: Seq[String] = Seq())(implicit ds: Datastore, futurePool: FuturePool): Future[Locality] = {
    futurePool {
      val query = ds.createQuery(classOf[Locality]).field(Locality.FD_ID).equal(id).retrievedFields(true, fields: _*)
      query.get()
    }
  }

  def getLocalityByIds(ids: Seq[ObjectId])(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Locality]] = {
    futurePool {
      if (ids == null || ids.isEmpty)
        Seq()
      else {
        val query = ds.createQuery(classOf[Locality]).field(Locality.FD_ID).in(ids).retrievedFields(true, Seq(Country.FN_ID, Locality.FD_ZH_NAME): _*)
        query.asList()
      }
    }
  }

  def getLocalityByCountryCode(countryId: ObjectId, fields: Seq[String] = Seq())(implicit ds: Datastore, futurePool: FuturePool): Future[Seq[Locality]] = {
    futurePool {
      val query = ds.createQuery(classOf[Locality]).field("country.id").equal(countryId).field("taoziEna").equal(true)
        .retrievedFields(true, fields: _*)
      query.asList()
    }
  }

}
